src/metadata.rs

```diff
@@ -0,0 +1,331 @@
+        ));
+    }
+    Ok(key)
+}
+
+/// Builder for immutable Metadata.
+pub struct MetadataBuilder {
+    arr: Metadata,
+}
+
+impl MetadataBuilder {
+    /// Create a builder with empty initial capacity.
+    pub fn new() -> MetadataBuilder {
+        MetadataBuilder::with_capacity(0)
+    }
+
+    /// Create a builder with the given value.
+    pub fn with_capacity(cap: usize) -> MetadataBuilder {
+        MetadataBuilder {
+            arr: Metadata::with_capacity(cap),
+        }
+    }
+
+    /// Add a metadata holding an ASCII value.
+    ///
+    /// `key` must not use suffix (-bin) indicating a binary valued metadata entry.
+    pub fn add_str(&mut self, key: &str, value: &str) -> Result<&mut MetadataBuilder> {
+        if !value.is_ascii() {
+            return Err(Error::InvalidMetadata(
+                "only ascii value is accepted.".to_owned(),
+            ));
+        }
+        for b in value.bytes() {
+            if 0 == unsafe { libc::isprint(b as i32) } {
+                return Err(Error::InvalidMetadata(
+                    "Only printable chars are accepted.".to_owned(),
+                ));
+            }
+        }
+        let key = normalize_key(key, false)?;
+        self.add_metadata(&key, value.as_bytes())
+    }
+
+    fn add_metadata(&mut self, key: &str, value: &[u8]) -> Result<&mut MetadataBuilder> {
+        unsafe {
+            grpc_sys::grpcwrap_metadata_array_add(
+                &mut self.arr.0,
+                key.as_ptr() as _,
+                key.len(),
+                value.as_ptr() as _,
+                value.len(),
+            )
+        }
+        Ok(self)
+    }
+
+    /// Add a metadata holding a binary value.
+    ///
+    /// `key` needs to have suffix (-bin) indicating a binary valued metadata entry.
+    pub fn add_bytes(&mut self, key: &str, value: &[u8]) -> Result<&mut MetadataBuilder> {
+        let key = normalize_key(key, true)?;
+        self.add_metadata(&key, value)
+    }
+
+    /// Create `Metadata` with configured entries.
+    pub fn build(mut self) -> Metadata {
+        unsafe {
+            grpc_sys::grpcwrap_metadata_array_freeze(&mut self.arr.0);
+        }
+        self.arr
+    }
+}
+
+/// A collection of metadata entries that can be exchanged during a call.
+///
+/// gRPC supports these types of metadata:
+///
+/// - Request headers
+///
+///     They are sent by the client at the beginning of a remote call before
+///     any request messages are sent.
+///
+/// - Response headers
+///
+///     They are sent by the server at the beginning of a remote call handler
+///     before any response messages are sent.
+///
+/// - Response trailers
+///
+///     They are sent by the server at the end of a remote call along with
+///     resulting call status.
+///
+/// Metadata value can be ascii string or bytes. They are distinguish by the
+/// key suffix, key of bytes value should have suffix '-bin'.
+#[repr(C)]
+pub struct Metadata(GrpcMetadataArray);
```

Comment:
I think it is better to name it `MetadataArray` to keep the naming consistency with gRPC, it is weird that `Metadata` actually contains an array of metadata.

Reply:
Actually its name is taken from [gRPC cshar](https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/Metadata.cs#L37).